lib/Graph/TensorLayout.cpp

```diff
@@ -0,0 +1,557 @@
+static std::string dims0D[]{
+    {""},
+};
+static std::string dims1D[] = {
+    {"N"},
+};
+static std::string dims2D[] = {
+    {"*"},
+    {"*"},
+};
+static std::string dims3D[] = {
+    {"*"},
+    {"*"},
+    {"*"},
+};
+static std::string dims4D[] = {
+    {"*"},
+    {"*"},
+    {"*"},
+    {"*"},
+};
+static std::string dims5D[] = {
+    {"*"}, {"*"}, {"*"}, {"*"}, {"*"},
+};
+static std::string dims6D[] = {
+    {"*"}, {"*"}, {"*"}, {"*"}, {"*"}, {"*"},
+};
+
+static TensorLayoutDescription layoutNHWC(dimsNHWC);
+static TensorLayoutDescription layoutNCHW(dimsNCHW);
+static TensorLayoutDescription layoutHWNC(dimsHWNC);
+static TensorLayoutDescription layout0D(dims0D);
+static TensorLayoutDescription layout1D(dims1D);
+static TensorLayoutDescription layout2D(dims2D);
+static TensorLayoutDescription layout3D(dims3D);
+static TensorLayoutDescription layout4D(dims4D);
+static TensorLayoutDescription layout5D(dims5D);
+static TensorLayoutDescription layout6D(dims6D);
+
+/// Glow layouts for any specific number of dimensions.
+static std::array<TensorLayoutDescription, max_tensor_dimensions + 1>
+    layoutsForDims = {
+        layout0D, layout1D, layout2D, layout3D, layout4D, layout5D, layout6D,
+};
+
+TensorLayoutCommon::TensorLayoutCommon() : enabled_(false) {
+  layoutNameToLayoutDescription_.insert(
+      std::make_pair("NCHW", new TensorLayoutDescription("NCHW")));
+  layoutNameToLayoutDescription_.insert(
+      std::make_pair("NHWC", new TensorLayoutDescription("NHWC")));
+  layoutNameToLayoutDescription_.insert(
+      std::make_pair("HWNC", new TensorLayoutDescription("HWNC")));
+  layoutNameToLayoutDescription_.insert(
+      std::make_pair("N", new TensorLayoutDescription("N")));
+}
+
+TensorLayoutCommon::~TensorLayoutCommon() {
+  while (!layoutNameToLayoutDescription_.empty()) {
+    auto curr = layoutNameToLayoutDescription_.begin();
+    auto *tld = curr->second;
+    layoutNameToLayoutDescription_.erase(curr);
+    delete tld;
+  }
+}
+
+std::array<TensorLayoutDescription, max_tensor_dimensions + 1> &
+TensorLayoutCommon::getLayoutsForDims() const {
+  return layoutsForDims;
+}
+
+static TensorLayoutDescription *
+getLayoutFromName(const std::string &name,
+                  std::unordered_map<std::string, TensorLayoutDescription *>
+                      &layoutNameToLayoutDescription) {
+  if (isAnyHelper(name)) {
+    return nullptr;
+  }
+  auto it = layoutNameToLayoutDescription.find(name);
+  if (it != layoutNameToLayoutDescription.end()) {
+    return it->second;
+  }
+  // Add new layout to map:
+  auto *ret = new TensorLayoutDescription(name);
+  layoutNameToLayoutDescription.insert(std::make_pair(name, ret));
+  return ret;
+}
+
+std::string TensorLayoutCommon::getDefaultNDLayout(unsigned dims) const {
+  DCHECK_LE(dims, max_tensor_dimensions) << "Too many dimensions";
+  return getLayoutsForDims()[dims].getSerializedLayout();
+}
+
+std::string TensorLayoutCommon::getNthInputLayoutRequirements(const Node *node,
```

Comment:
Could you add a doxygen comment about a general logic of such a function? Something like "Certain nodes are layout agnostic. Others expect their inputs/outputs to have a canonical format. For some layout agnostic nodes we need to look at the layout of their inputs to determine the layout of their outputs, e.g. ... "

Reply:
Done - added that to the header file.